packages/story-editor/src/utils/useIdleTaskQueue.js

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import { useRef, useCallback } from '@googleforcreators/react';
+
+/**
+ * Internal dependencies
+ */
+import { requestIdleCallback, cancelIdleCallback } from './idleCallback';
+
+/**
+ * Creates a FIFO idle task queue
+ *
+ * @return {Function} queueIdleTask
+ */
+function useIdleTaskQueue() {
+  const taskQueue = useRef([]);
+  const isTaskQueueRunning = useRef(false);
+  const currentTask = useRef([null, null]);
+
+  /**
+   * Recursively runs idle task queue sequentially
+   * until all queued tasks are run.
+   *
+   * @return {void}
+   */
+  const runTaskQueue = useCallback(() => {
+    isTaskQueueRunning.current = true;
+
+    if (!taskQueue.current.length) {
+      isTaskQueueRunning.current = false;
+      return;
+    }
+
+    const [taskId, task] = taskQueue.current.shift();
+    const idleCallbackId = requestIdleCallback(async () => {
+      await task();
+      currentTask.current = [null, null];
+      runTaskQueue();
+    });
+    currentTask.current = [taskId, idleCallbackId];
+  }, []);
```

Comment:
Cool to see that we have a fallback in `requestIdleCallback` for safari users too: https://github.com/GoogleForCreators/web-stories-wp/blob/c0a81152b0174c7daa8ff6f6bc1f93047d129431/packages/story-editor/src/utils/idleCallback.js#L18-L37